def feature_counts_dynamic(name, manager):

    metadata = {'name': name}
    feature_counts_code = '''

rule feature_counts_%(name)s:
    """
    Feature counts (subread) summarizes counts mapped reads for 
    genomic features such as genes, exons, promoter, gene bodies, 
    genomic bins and chromosomal locations.

    :reference: http://bioinf.wehi.edu.au/featureCounts/

    Required input:
         __feature_counts__input: sorted bam file

    Required output:
        __feature_counts__output_count: output tabulated-delimited file

    Config:

        .. code-block:: yaml

            feature_counts:
                gff: " "       #path to the GFF/GTF annotation file
                options:  " "  #options for featureCounts you want use
    """


    input:
        bam = __feature_counts__input,
        gff = __feature_counts__gff
    output:
        count = __feature_counts_%(name)s__output_count
    params:
        mapp = __feature_counts__options
    log:
        __feature_counts_%(name)s__log
    threads:
        __feature_counts__threads
    run:
        shell("""featureCounts -T {threads} {params.mapp} \
                 -a {input.gff} -o {output.count} {input.bam} 2> {log}""")
        # now we can read the summary file back and add a tag
        sample = "__fc{}_".format(str(%(name)s)) + input.bam.split("/")[0]
        cmd = 'sed -i "s/Status.*/Status\t{}/g" {}.summary'.format(sample, output.count)
        shell(cmd)

    '''
    from sequana.snaketools import build_dynamic_rule
    return build_dynamic_rule(feature_counts_code % metadata, manager.pipeline_dir)
